File: app/src/main/java/com/example/schooldiary/ui/bottomNavigation/mainfragments/NewsFragment.java
package com.example.schooldiary.ui.bottomNavigation.mainfragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.schooldiary.R;
import com.example.schooldiary.model.News;
import com.example.schooldiary.ui.adapters.firestorerecycler.FirestoreRecyclerAdapter;
import com.example.schooldiary.ui.adapters.firestorerecycler.FirestoreRecyclerOptions;
import com.example.schooldiary.ui.adapters.news.NewsFragmentAdapter;
import com.example.schooldiary.ui.adapters.news.NewsHolder;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;

public class NewsFragment extends Fragment implements NewsHolder.NewsClickListener {

    private RecyclerView newsRecyclerView;
    private FirestoreRecyclerAdapter adapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_news, container, false);
        newsRecyclerView = root.findViewById(R.id.news_recycler_view);
        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {

        FirebaseFirestore firestore = FirebaseFirestore.getInstance();
        Query query = firestore.collection("news")
                .orderBy("serialNumber", Query.Direction.DESCENDING);

        FirestoreRecyclerOptions<News> options = new FirestoreRecyclerOptions.Builder<News>()
                .setQuery(query, News.class)
                .build();

        adapter = new NewsFragmentAdapter(options, this);

        newsRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        newsRecyclerView.setAdapter(adapter);
    }

    @Override
    public void onStart() {
        super.onStart();
        adapter.startListening();
    }

    @Override
    public void onStop() {
        super.onStop();
        adapter.stopListening();
    }

    @Override
    public void openFragment(int id, String photo, String title, String text) {
        Bundle bundle = new Bundle();
        bundle.putString("Photo", photo);
        bundle.putString("Title", title);
        bundle.putString("Text", text);
        NavController navController = Navigation.findNavController(getActivity(), R.id.nav_host_fragment);
        navController.navigate(id, bundle);
    }
}

File: app/src/main/java/com/example/schooldiary/ui/bottomnavigation/mainfragments/NewsFragment.kt
package com.example.schooldiary.ui.bottomnavigation.mainfragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.schooldiary.R
import com.example.schooldiary.databinding.FragmentNewsBinding
import com.example.schooldiary.model.news.News
import com.example.schooldiary.ui.adapters.news.NewsFragmentAdapter
import com.example.schooldiary.ui.adapters.news.NewsHolder
import com.example.schooldiary.ui.adapters.news.NewsHolder.NewsClickListener
import com.firebase.ui.firestore.FirestoreRecyclerAdapter
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query

class NewsFragment : Fragment(), NewsClickListener {
    private var _binding: FragmentNewsBinding? = null
    private val binding get() = _binding!!
    private lateinit var adapter: FirestoreRecyclerAdapter<News, NewsHolder>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        _binding = FragmentNewsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val firestore = FirebaseFirestore.getInstance()
        val query = firestore.collection(FIREBASE_NEWS_COLLECTION)
                .orderBy(FIREBASE_SERIAL_NUMBER_FIELD, Query.Direction.DESCENDING)
        val options = FirestoreRecyclerOptions.Builder<News>()
                .setQuery(query, News::class.java)
                .build()
        adapter = NewsFragmentAdapter(options, this)

        binding.recyclerView.let {
            it.adapter = adapter
            it.layoutManager = LinearLayoutManager(activity)
        }
    }

    override fun onStart() {
        super.onStart()
        adapter.startListening()
    }

    override fun onStop() {
        super.onStop()
        adapter.stopListening()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun openFragment(id: Int, photo: String, time: String, title: String, text: String) {
        val bundle = bundleOf(ARG_NEWS to News(photo, time, title, text))
        activity?.let {
            val navController = Navigation.findNavController(it, R.id.nav_host_fragment)
            navController.navigate(id, bundle)
        }
    }

    companion object {
        private const val FIREBASE_SERIAL_NUMBER_FIELD = "serialNumber"
        private const val FIREBASE_NEWS_COLLECTION = "news"
        private const val ARG_NEWS = "news"
    }
}